omod/src/main/web/dashboardwidgets/programstatus/programstatus.controller.js

```diff
@@ -466,8 +463,33 @@ export default class ProgramStatusController {
             this.patientProgram.states = this.$filter('filter')(this.patientProgram.states, (state) => {
                 return !state.voided
             }, true);
-            this.patientProgram.states = this.$filter('orderBy')(this.patientProgram.states, (state) => {
-                return new Date(state.startDate);
+            this.patientProgram.states = this.$filter('orderBy')(this.patientProgram.states, null, false, function(state1, state2)
+            {
+                //orderBy falls back to index comparison when two states are indicated to be equal
+                if(state1.type === "number" && state2.type === "number"){
+                    //if index of state1 is 7 and state2 is 8, 7-8 == -1 indicates state1 is first, 8-7 == 1 indicating state2 is first
+                    return state1.value-state2.value;
+                }
```

Comment:
Why did we end up adding a comparsion by index here?  Are we relying on the service returning the states ordered by DB primary key (when all else is equal)?  We should never be relying on index because although in the general use case earlier created states have lower primary keys, we make no guarantees, for example in term of migrations, synced databases, etc.

Reply:
This is actually an angular behavior (and relates to the index of the element in the array being sorted)
```
In order to ensure that the sorting will be deterministic across platforms, if none of the specified predicates can distinguish between two items, orderBy will automatically introduce a dummy predicate that returns the item's index as value. (If you are using a custom comparator, make sure it can handle this predicate as well.)
```
- https://docs.angularjs.org/api/ng/filter/orderBy

Reply:
Wow, interesting... still not sure I 100% understand, though. :)

Can we extract this into an util method and through some test cases around it?

Reply:
Which part? The orderby comparator? 

>  still not sure I 100% understand, though. :)

I was seeing it in the test. There are two program workflows, so it's possible to have two null end dates. Firefox was tending to sort two states in reverse order vs. chrome.

If the comparator says the two states are equal ( return 0; ), the two states are re-evaluated with the same comparator by their array index to provide a stable sort (the order is always the same given the same "verbatim" input, vs. an unstable sort where any two "equal" states could be in either order given that same input, generally this is seen as a symptom of quick sort pivot randomness).

Reply:
Ah, I thought you had added test cases, but then I didn't see them when I last checked because Github had collapsed that file because it was large... I'll check out the changes now.

Reply:
What confuses me about this is the fact that the fallback test appears first... it is that when orderBy calls the custom comparator and the comparator returns 0 it then converts the value to a number based on index and calls the comparator again?

Reply:
> it is that when orderBy calls the custom comparator and the comparator returns 0 it then converts the value to a number based on index and calls the comparator again?

Yes (pretty much). It doesn't "convert to a number" so much as just fallback to passing the index instead of the obj. If you're not familiar with JS array prototypes, most (all?), like Array.prototype.forEach, can optionally receive the index, 
```
[3,2,1].forEach( function(obj, index) { console.log("array[",index,"]=",obj) } );

array[ 0 ]= 3
array[ 1 ]= 2
array[ 2 ]= 1
```

> What confuses me about this is the fact that the fallback test appears first...

I added it first as the "base case" so the rest of the code after it doesn't have to check for stateX.type="number" or vice versa. This is essentially the "type checking" code.

JS is weakly typed, and Angular converts this into it's own object as described in the [orderby doc](https://docs.angularjs.org/api/ng/filter/orderBy). Even in Java it wouldn't be as simple as just having two overloaded function signatures on a comparator (though that would be nice), because the ABC requires the Object function signature be overriden and implemented (which is where at least the type checking would occur, but could call corresponding functions for a given type instead of doing all the comparisons in the one function).

If you prefer these "two" comparator logic blocks (one for "number" and one for the state "object") be made into functions and called based on the "type checking" for clarity, I can accommodate.

Reply:
No, that's fine, works for me!  thanks for the detailed description!  Perhaps just add a comment to the text either linking to this discussion or some other explanation would be helpful, because I think others may be confused like me.

Reply:
Ok, Thanks!